contractcourt/anchor_resolver.go

```diff
@@ -0,0 +1,167 @@
+package contractcourt
+
+import (
+	"errors"
+	"io"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/lnwallet"
+	"github.com/lightningnetwork/lnd/sweep"
+)
+
+// anchorResolver is a resolver that will attempt to sweep our anchor output.
+type anchorResolver struct {
+	// anchorSignDescriptor contains the information that is required to
+	// sweep the anchor.
+	anchorSignDescriptor input.SignDescriptor
+
+	// anchor is the outpoint on the commitment transaction.
+	anchor wire.OutPoint
+
+	// resolved reflects if the contract has been fully resolved or not.
+	resolved bool
+
+	// broadcastHeight is the height that the original contract was
+	// broadcast to the main-chain at. We'll use this value to bound any
+	// historical queries to the chain for spends/confirmations.
+	broadcastHeight uint32
+
+	// chanPoint is the channel point of the original contract.
+	chanPoint wire.OutPoint
+
+	contractResolverKit
+}
+
+// newAnchorResolver instantiates a new anchor resolver.
+func newAnchorResolver(anchorSignDescriptor input.SignDescriptor,
+	anchor wire.OutPoint, broadcastHeight uint32,
+	chanPoint wire.OutPoint, resCfg ResolverConfig) *anchorResolver {
+
+	r := &anchorResolver{
+		contractResolverKit:  *newContractResolverKit(resCfg),
+		anchorSignDescriptor: anchorSignDescriptor,
+		anchor:               anchor,
+		broadcastHeight:      broadcastHeight,
+		chanPoint:            chanPoint,
+	}
+
+	r.initLogger(r)
+
+	return r
+}
+
+// ResolverKey returns an identifier which should be globally unique for this
+// particular resolver within the chain the original contract resides within.
+func (c *anchorResolver) ResolverKey() []byte {
+	// The anchor resolver is stateless and doesn't need a database key.
+
+	return nil
+}
+
+// Resolve offers the anchor output to the sweeper and waits for it to be swept.
+//
+// NOTE: This function MUST be run as a goroutine.
+func (c *anchorResolver) Resolve() (ContractResolver, error) {
+	// Attempt to update the sweep parameters to the post-confirmation
+	// situation. We don't want to force sweep anymore, because the anchor
+	// lost its special purpose to get the commitment confirmed. It is just
+	// an output that we want to sweep only if it is economical to do so.
+	relayFeeRate := c.Sweeper.RelayFeePerKW()
+
+	resultChan, err := c.Sweeper.UpdateParams(
+		c.anchor,
+		sweep.ParamsUpdate{
+			Fee: sweep.FeePreference{
+				FeeRate: relayFeeRate,
+			},
+		},
+	)
+
+	// After a restart or when the remote force closes, the sweeper is not
+	// yet aware of the anchor. In that case, offer it as a new input to the
+	// sweeper. An exclusive group is not necessary anymore, because we know
+	// that this is the only anchor that can be swept.
+	if err == lnwallet.ErrNotMine {
+		anchorInput := input.MakeBaseInput(
```

Comment:
can we just offer it like this in both cases, or won't that override what is already in the sweeper?

Reply:
Yes, won't override indeed.